W-14604001 java17 landing page vh (#201)
mulesoft/docs-general: modules/ROOT/pages/upgrade-policies-proxies.adoc

= Upgrade Java for Policies and API Proxies 

Upgrade your policies and API proxies to be compatible with Java 17.

[[upgrading-automated-policies]]
== Upgrading Automated Policies

Upgrade your automated policies (included and custom) to Java 17.

[[upgrading-included-automated-policies]]
=== Upgrading Included Automated Policies

To upgrade automated xref:mule-gateway::policies-included-directory.adoc[MuleSoft included policies] to Java 17:

. In API Manager, go to *Automated Policies*.
. In *Policy version*, select the latest version of the included policy.
. In *Rule of application*:
  * To apply included policies to all Mule runtimes, select *All runtimes* and click *Apply*.
  * To apply the included policies to Flex Gateways only, select *Flex Gateways only* and click *Apply*.

>   * To apply the included policies to Mule Gateways only:

Reviewer: Does the indentation work? Never tried this, please confirm in a local build. Otherwise, use **, but if this works, great!

>   ** Select *Mule Gateways only*.
>   ** In *From version* and *To*, enter the Mule version range.
>   ** In *Java versions*, select *All Java versions* or *Choose specific Java versions*. 
>   ** Click *Apply*.

Reviewer: ```suggestion
  .. Select *Mule Gateways only*.
  .. In *From version* and *To*, enter the Mule version range.
  .. In *Java versions*, select *All Java versions* or *Choose specific Java versions*. 
  .. Click *Apply*.
```
Changing to ordered list because these are ordered steps again. Please confirm if everything renders as intended after the changes.

[[upgrading-custom-automated-policies]]
=== Upgrading Custom Automated Policies

To upgrade custom automated policies to Java 17:

. In API Manager, go to *Automated Policies*.
. Upgrade the required extensions and connectors to Java 17.
 * If your custom policy has Anypoint Connectors, see xref:java-support.adoc#anypoint-connectors-and-modules[Upgrading Anypoint Connectors and Modules].

>  * If your custom policy has custom connectors, see <<Upgrading Custom Connectors INSERT LINK>>.

Reviewer: The INSERT LINK here has to be replaced with the proper anchor right?

. xref:gateway::policies-custom-java-version.adoc[Upgrade the custom policy (link to new page in Gateway)].
. xref:gateway::policies-custom-package.adoc[Package the custom policy].
. xref:gateway::policies-custom-upload-to-exchange.adoc[Republish the custom policy to Exchange].
. Apply the new version of the custom policy that is compatible with Java 17 to the API instance:
 .. Go to *Policies*.
 .. In *Applied policy status*, verify that the latest definition is applied.
 * If the latest definition is applied, click on the More (...) menu and select *Check for implementation updates*.
 * If the latest definition is not applied, update to the latest definition.

[[upgrading-api-level-policies]]
== Upgrading API-Level Policies

Upgrade your API-level policies (included and custom) to Java 17.

[[upgrading-included-api-level-policies]]
=== Upgrading Included API-Level Policies

To upgrade API-level xref:mule-gateway::policies-included-directory.adoc[MuleSoft included policies] to Java 17:

> . In API Manager, go to *API Administration > <API Name> > Policies > Add a policy* to xref:gateway::policies-included-apply.adoc[apply policies] at the individual API level. 

Reviewer: Wondering if there's any alternative to **<API Name>** because the angle brackets might get mixed with the > indicating next. I think asciidoc renders an arrow if you use --> 
Otherwise, maybe use curly brackets here? to: *API Administration > {API Name} > Policies > Add a policy*
I'll leave you to decide what's best in this situation. Making no change is also a valid option.

. Upgrade each of your API-level policies:
.. In *Policy version*, select the latest version of the included policy.
.. In *Method & resource conditions*, select either *Apply configuration to all API method & resources* or *Apply configuration to specific API method & resources*.
.. Click *Apply*.

[[upgrading-custom-api-level-policies]]
=== Upgrading Custom API-level Policies

To upgrade custom API-level policies to Java 17:

> . In API Manager, go to *API Administration > <API Name> > Policies > Add a policy* to xref:gateway::policies-included-apply.adoc[apply policies] at the individual API level. 

Reviewer: Same comment about <API Name>

. Upgrade the required extensions and connectors to Java 17.
 * If your custom policy has Anypoint Connectors, see xref:java-support.adoc#anypoint-connectors-and-modules[Upgrading Anypoint Connectors and Modules].

>  * If your custom policy has custom connectors, see <<Upgrading Custom Connectors INSERT LINK>>.

Reviewer: Same comment about INSERT LINK

. xref:gateway::policies-custom-upgrade.adoc[Upgrade the custom policy (link to new page in Gateway)].
. xref:gateway::policies-custom-package.adoc[Package the custom policy].
. xref:gateway::policies-custom-upload-to-exchange.adoc[Republish the custom policy to Exchange].
. Apply the new version of the custom policy that is compatible with Java 17 to the API instance:
 .. Go to *Policies*.
 .. In *Applied policy status*, verify that the latest definition is applied.
 * If the latest definition is applied, click on the More (...) menu and select *Check for implementation updates*.
 * If the latest definition is not applied, update to the latest definition.

[[upgrading-api-proxies]]
== Upgrading API Proxies

[IMPORTANT]
To ensure your API proxies or Mule apps are protected when upgrading, upgrade your API policies *before* upgrading your API proxies or Mule apps.

> The steps to upgrade API proxies to Java 17 are a little different, depending on which deployment model you use. 

Reviewer: ```suggestion
The steps to upgrade API proxies to Java 17 differ depending on which deployment model you use. 
```

* If you use a Basic endpoint API instance to update your instance, update the Mule application connecting to your API instance.

[[cloudhub-and-cloudhub2-deployments]]
=== CloudHub and CloudHub 2.0 Deployments

Before following these steps, ensure your policies are updated. For more information, see <<upgrading-automated-policies>> and <<upgrading-api-level-policies>>. 

. In API Manager, go to *API Administration > Settings*.
. xref:api-manager::edit-api-endpoint-task.adoc[Edit the API instance's configuration settings]:  
 .. In *Runtime Channel*, select the runtime channel to use.
 .. In *Version*, select *4.6.0*.
 .. In *Java version*, select *Java 17*.
 . Click *Save & Apply*.

[[hybrid-deployments]]
=== Hybrid Deployments

Before following these steps, ensure your policies are updated. For more information, see <<upgrading-automated-policies>> and <<upgrading-api-level-policies>>. 

. Deploy Mule runtime engine 4.6 in a new server running on Java 17 and start the server. For more information, see xref:runtime-manager::servers-create.adoc[].

> . Select the new target you created running on Java 17 and in API Manager, go to *API Administration > Settings*.

Reviewer: ```suggestion
. Select the new target you created running on Java 17 and, in API Manager, go to *API Administration > Settings*.
```

. xref:api-manager::edit-api-endpoint-task.adoc[Edit the API instance's configuration settings]:  
 .. In *Select target*, select the server running on Java 17 as the target.
. Click *Save & Apply*.

> . After the server successfully runs on Java 17, shift your traffic gradually to the new server using load balancer and turn off the old server after all of its apps are migrated.

Reviewer: ```suggestion
. After the server successfully runs on Java 17, shift your traffic gradually to the new server using a load balancer and turn off the old server after all of its apps are migrated.
```

[[runtime-fabric-deployments]]
=== Runtime Fabric Deployments 

Before following these steps, ensure your policies are updated. For more information, see <<upgrading-automated-policies>> and <<upgrading-api-level-policies>>. 

. Deploy Mule runtime engine 4.6 in a new server running on Java 17 and start the server. For more information, see xref:runtime-fabric::proxy-deploy-runtime-fabric.adoc[].

> . Select the new target you created running on Java 17 and in API Manager, go to *API Administration > Settings*.

Reviewer: ```suggestion
. Select the new target you created running on Java 17 and, in API Manager, go to *API Administration > Settings*.
```

. xref:api-manager::edit-api-endpoint-task.adoc[Edit the API instance's configuration settings]:  
 .. In *Runtime Channel*, select the runtime channel to use.
 .. In *Version*, select *4.6.0*.
 .. In *Java version*, select *Java 17*.
. Click *Save & Apply*.

> . After the server successfully runs on Java 17, shift your traffic gradually to the new server using load balancer and turn off the old server after all of its apps are migrated.

Reviewer: ```suggestion
. After the server successfully runs on Java 17, shift your traffic gradually to the new server using a load balancer and turn off the old server after all of its apps are migrated.
```
